File: hard/01 - LCS Dynamic-Programming/01.js
function longestCommonSubsequence(str1, str2) {
    if (!str1 || !str2) {
      return [];
    }
  
    const r = str1.length;
    const c = str2.length;
    const dp = Array(c + 1).fill('');
  
    for (let row = 1; row <= r; row++) {
      let prev = dp[0];
      for (let col = 1; col <= c; col++) {
        const temp = dp[col];
        if (str1[row - 1] === str2[col - 1]) {
          dp[col] = prev + str1[row - 1];
        } else {
          dp[col] = [dp[col], dp[col - 1], prev].sort((a, b) => b.length - a.length)[0];
        }
        prev = temp;
      }
    }
  
    return dp[c].split('');
  }
  
  

File: hard/01 - LCS Dynamic-Programming/01.py
def longest_common_subsequence(str1, str2):
    if not str1 or not str2:
        return []

    r, c = len(str1), len(str2)
    dp = [""] * (c + 1)

    for row in range(1, r + 1):
        prev = dp[0]
        for col in range(1, c + 1):
            temp = dp[col]
            if str1[row - 1] == str2[col - 1]:
                dp[col] = prev + str1[row - 1]
            else:
                dp[col] = max(dp[col], dp[col - 1], key=len)
            prev = temp

    return list(dp[c])
